lib/common/redis.c

```diff
@@ -90,73 +151,267 @@ void h2o_redis_connect(h2o_redis_conn_t *conn, const char *host, uint16_t port)
 
     if (redis->err != REDIS_OK) {
         /* some connection failures can be detected at this time */
-        conn->_timeout_entry.cb = on_connect_error_deferred;
-        h2o_timeout_link(conn->loop, &conn->_defer_timeout, &conn->_timeout_entry);
+        invoke_deferred(conn, &conn->_defer_timeout_entry, on_connect_error_deferred);
         return;
     }
 
+    conn->_did_connect_timeout = 0;
+    if (conn->connect_timeout != 0) {
+        h2o_timeout_init(conn->loop, &conn->_connect_timeout, conn->connect_timeout);
+        h2o_timeout_link(conn->loop, &conn->_connect_timeout, &conn->_connect_timeout_entry);
+    }
+
     attach_loop(redis, conn->loop);
-    redisAsyncSetConnectCallback(redis, on_redis_connect);
-    redisAsyncSetDisconnectCallback(redis, on_redis_disconnect);
+    redisAsyncSetConnectCallback(redis, on_connect);
+    redisAsyncSetDisconnectCallback(redis, on_disconnect);
 }
 
 void h2o_redis_disconnect(h2o_redis_conn_t *conn)
 {
-    if (conn->state != H2O_REDIS_CONNECTION_STATE_CLOSED) {
-        assert(conn->_redis != NULL);
-        conn->state = H2O_REDIS_CONNECTION_STATE_CLOSED;
-        redisAsyncDisconnect(conn->_redis);
-    }
+    disconnect(conn, 0);
 }
 
-static void on_command(redisAsyncContext *redis, void *reply, void *privdata)
+static void dispose_command(h2o_redis_command_t *command)
 {
-    struct st_h2o_redis_command_t *command = (struct st_h2o_redis_command_t *)privdata;
-    if (command->cb != NULL) {
-        command->cb((redisReply *)reply, command->data);
+    if (h2o_timeout_is_linked(&command->_defer_timeout_entry)) {
+        h2o_timeout_unlink(&command->_defer_timeout_entry);
     }
+
+    if (h2o_timeout_is_linked(&command->_command_timeout_entry)) {
+        h2o_timeout_unlink(&command->_command_timeout_entry);
+    }
+    if (h2o_linklist_is_linked(&command->_command_timeout._link)) {
+        h2o_timeout_dispose(command->conn->loop, &command->_command_timeout);
+    }
+
     free(command);
 }
 
+static void return_reply(redisAsyncContext *redis, redisReply *reply, h2o_redis_command_t *command)
+{
+    if (command->cb == NULL)
+        return;
+
+    int err = H2O_REDIS_ERROR_NONE;
+    char *errstr = NULL;
```

Comment:
Do we need `err` and `errstr`?

I would appreciate it if you could merge them instead of having two. One way would be to only use `const char *` and have specific pointer values defined in the header file. Please consult http1client for details.

Reply:
At https://github.com/h2o/h2o/commit/3ead0dac3b178453c9b592aa3b86fffb3f42a16b I remoted `int err`, but still I'm feeling that it may be necessary to report better error message to the client, regarding the fact that error message may came from `strerror`